### PR TITLE
Fix #230, ability to choose different search engine per browsing mode.

### LINKED ISSUE #230: Normal vs Private Search Engine Settings
Currently there is only a single "default search engine". We need to provide the dual search engine support that 1.6 currently has. A default search engine for both normal and private mode.

### PR DESCRIPTION
**Pull request by iccub**
Code was mostly copied from 1.6 with minor modifications.
All search related classes were moved to a separate folder.

## Pull Request Checklist

- [ ] My patch has gone through review and I have addressed review comments
- [x] My patch has a standard commit message that looks like `Fix #123: This fixes the shattered coffee cup!`

- [ ] I have updated the *Unit Tests* to cover new or changed functionality
- [ ] I have updated the *UI Tests* to cover new or changed functionality
- [ ] I have marked the bug with `[needsuplift]`
- [ ] I have made sure that localizable strings use `NSLocalizableString()`

## Screenshots

If your patch includes user interface changes that you would like to suggest or that you would like UX to look at, please include them here.

## Notes for testing this patch

If useful, please leave notes for QA, explaining what this patch changes and how it can be best tested and verified.

### CODE REVIEW

**Comment by iccub** [comment]
Changing to [WIP] as some tests are still broken.

**Comment by iccub** [comment]
Tests are working now.

File: Client/Frontend/Browser/Search/SearchEngines.swift [hunk 1]
```diff
@@ -15,6 +15,12 @@ private let ShowSearchSuggestionsOptIn = "search.suggestions.showOptIn"
 private let ShowSearchSuggestions = "search.suggestions.show"
 private let customSearchEnginesFileName = "customEngines.plist"
 
+// BRAVE TODO: Move to newer Preferences class.
```
Comment (kylehickinson) [inline Client/Frontend/Browser/Search/SearchEngines.swift]: Can you please make a ticket for this, put it under the Preferences epic and then update this TODO to include the ticket number.

File: Client/Frontend/Browser/Search/SearchEngines.swift [hunk 2]
```diff
@@ -45,24 +53,33 @@ class SearchEngines {
         self.disabledEngineNames = getDisabledEngineNames()
         self.orderedEngines = getOrderedEngines()
     }
-
-    var defaultEngine: OpenSearchEngine {
-        get {
+    
+    /// If no engine type is specified this method returns search engine for regular browsing.
+    func defaultEngine(forType type: DefaultEngineType? = nil) -> OpenSearchEngine {
+            let engineType = type ?? (PrivateBrowsingManager.shared.isPrivateBrowsing ? .privateMode : .standard)
```
Comment (kylehickinson) [inline Client/Frontend/Browser/Search/SearchEngines.swift]: Accidental indentation?

File: Client/Frontend/Browser/Search/SearchEngines.swift [hunk 3]
```diff
@@ -45,24 +53,33 @@ class SearchEngines {
...
-            // The default engine is always first in the list.
-            var orderedEngines = self.orderedEngines.filter { engine in engine.shortName != defaultEngine.shortName }
-            orderedEngines.insert(defaultEngine, at: 0)
-            self.orderedEngines = orderedEngines
-        }
+    func setDefaultEngine(_ engine: String, forType type: DefaultEngineType) {
+        prefs.setString(engine, forKey: type.rawValue)
+        
+        // The default engine is always enabled.
+        self.enableEngine(defaultEngine(forType: type))
```
Comment (kylehickinson) [inline Client/Frontend/Browser/Search/SearchEngines.swift]: Don't need `self`, here and the few lines below

File: Client/Frontend/Settings/SearchSettingsTableViewController.swift [hunk 4]
```diff
@@ -153,18 +162,27 @@ class SearchSettingsTableViewController: UITableViewController {
...
         }
     }
 
     override func tableView(_ tableView: UITableView, willSelectRowAt indexPath: IndexPath) -> IndexPath? {
         if indexPath.section == SectionDefault && indexPath.item == ItemDefaultEngine {
-            let searchEnginePicker = SearchEnginePicker()
+            let searchEnginePicker = SearchEnginePicker(type: .standard)
+            // Order alphabetically, so that picker is always consistently ordered.
+            // Every engine is a valid choice for the default engine, even the current default engine.
+            searchEnginePicker.engines = model.orderedEngines.sorted { e, f in e.shortName < f.shortName }
```
Comment (kylehickinson) [inline Client/Frontend/Settings/SearchSettingsTableViewController.swift]: You can probably omit argument names here: 
```swift
model.orderedEngines.sorted { $0.shortName < $1.shortName }
```

File: Client/Frontend/Settings/SearchSettingsTableViewController.swift [hunk 5]
```diff
@@ -153,18 +162,27 @@ class SearchSettingsTableViewController: UITableViewController {
...
+            searchEnginePicker.delegate = self
+            searchEnginePicker.selectedSearchEngineName = model.defaultEngine(forType: .standard).shortName
+            navigationController?.pushViewController(searchEnginePicker, animated: true)
+        }
+        else if indexPath.section == SectionDefault && indexPath.item == ItemDefaultPrivateEngine {
+            let searchEnginePicker = SearchEnginePicker(type: .privateMode)
             // Order alphabetically, so that picker is always consistently ordered.
             // Every engine is a valid choice for the default engine, even the current default engine.
             searchEnginePicker.engines = model.orderedEngines.sorted { e, f in e.shortName < f.shortName }
```
Comment (kylehickinson) [inline Client/Frontend/Settings/SearchSettingsTableViewController.swift]: Same as above

File: ClientTests/SearchEnginesTests.swift [hunk 6]
```diff
@@ -28,7 +28,7 @@ class SearchEnginesTests: XCTestCase {
         // If this is our first run, Google should be first for the en locale.
         let profile = MockProfile()
         let engines = SearchEngines(prefs: profile.prefs, files: profile.files)
-        XCTAssertEqual(engines.defaultEngine.shortName, DefaultSearchEngineName)
+        XCTAssertEqual(engines.defaultEngine().shortName, DefaultSearchEngineName)
```
Comment (kylehickinson) [inline ClientTests/SearchEnginesTests.swift:31]: Should we add unit tests for private search engines too?

**Review by kylehickinson — CHANGES_REQUESTED** [review]

File: ClientTests/SearchEnginesTests.swift (hunk 6 above)
Comment (iccub) [reply ClientTests/SearchEnginesTests.swift:31]: Good idea, I added one test and one more assertion to verify that default private engine is set.

**Review by kylehickinson — DISMISSED** [review]

**Comment by kylehickinson** [comment]
Just found a bug after approval while still fiddling around: Changing the private tab search engine (when they are still defaulted both to Google, i.e. no changes made to either yet) changed the normal tab search engine as well.

Not sure if this is intended behaviour, can you please clarify? @iccub

**Comment by iccub** [comment]
Definitely a bug

**Review by kylehickinson — APPROVED** [review]

**Pull request merged**